test/e2e/kubectl/kubectl.go

```diff
@@ -514,8 +514,28 @@ var _ = SIGDescribe("Kubectl client", func() {
 			runOutput := framework.NewKubectlCommand(nsFlag, "run", "run-test", "--image="+busyboxImage, "--restart=OnFailure", "--attach=true", "--stdin", "--", "sh", "-c", "cat && echo 'stdin closed'").
 				WithStdinData("abcd1234").
 				ExecOrDie()
-			Expect(runOutput).To(ContainSubstring("abcd1234"))
-			Expect(runOutput).To(ContainSubstring("stdin closed"))
+
+			g := func(pods []*v1.Pod) sort.Interface { return sort.Reverse(controller.ActivePods(pods)) }
+			runTestPod, _, err := polymorphichelpers.GetFirstPod(f.ClientSet.CoreV1(), ns, "run=run-test", 1*time.Minute, g)
+			if err != nil {
+				os.Exit(1)
+			}
+			// NOTE: we cannot guarantee our output showed up in the container logs before stdin was closed, so we have
+			// to loop test.
+			err = wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
+				if !framework.CheckPodsRunningReady(c, ns, []string{runTestPod.Name}, 1*time.Second) {
+					framework.Failf("Pod %q of Job %q should still be running", runTestPod.Name, "run-test")
+				}
+				logOutput := framework.RunKubectlOrDie(nsFlag, "logs", runTestPod.Name)
+				Expect(runOutput).To(ContainSubstring("abcd1234"))
+				Expect(runOutput).To(ContainSubstring("stdin closed"))
+				return strings.Contains(logOutput, "abcd1234"), nil
+			})
+			if err != nil {
+				os.Exit(1)
+			}
+			Expect(err).To(BeNil())
```

Comment:
line 534-536 seems unnecessary because of line 537.

Reply:
Agreed and fixed. I had copied some boilerplate from the check the at the end of the test which was already using kubectl logs for similar reasons. I went ahead and cleaned up those redundant checks as well.

Reply:
Nice catch, thanks for updating